### ISSUE TITLE
SQL not compliant with SQL*Plus or SQL Developer

### ISSUE THREAD

**Issue by outon** [antlr/grammars-v4#1635]
I don't know if this can be considered an issue or a suggestion...

In current grammar I may have

```
SELECT 1 FROM DUAL
SELECT 2 FROM DUAL
```
and this is considered as two different queries when using "sql_script".
```
sql_script
    unit_statement:47
        data_manipulation_language_statements:3
            select_statement
    unit_statement:47
        data_manipulation_language_statements:3
            select_statement
    <EOF>
```
But this script although correct according to the grammar cannot be executed in **SQL*Plus** or **SQL Developer** because a semicolon is needed after each statement therefore is not considered valid.

The same goes for procedures, functions, triggers, anonymous blocks. The grammar considers that they end in the `END;` but in SQL*Plus you need a "/" in the next line to execute it.

Can this be considered as an issue?

**Comment by KvanTTT** [comment]
What grammar do you mean? T-Sql, PL/SQL or MySql?

**Comment by outon** [comment]
Sorry, I forgot to mention that I was talking about PL/SQL

**Comment by Pilo11** [comment]
This is a bit more difficult. The SQL*Plus is more restrictive than the SQL Developer.

If you try to execute: 
`insert into A (ID,TEST) values ('0', 'test text');
/`

with the SQL Developer: it's all fine. You will see a new wonderful entry in table 'A'.
with SQL*Plus or SQLcl: Error UNIQUE constraint violated (Column ID, primary key). Because the insert command will be written into the SQL buffer and the Slash will execute the SQL buffer, so in SQL*Plus it will be executed twice...

So you have to differentiate for whom the grammar is made. For the SQL Developer syntax checker? For the SQLcl/SQL*Plus script executer?

If you want to execute command with an ODBC/JDBC driver you have to execute every single command (no scripts possible):
If you execute: "ALTER TABLE A drop COLUMN TEST;" it will throw an exception (ora-01735)
If you execute the same statement without the semicolon it will work...

**Comment by outon** [comment]
I agree with you. Different programs can lead to different ways of interpreting the same grammar. It shoudn't, but it happens.

So we should ask ourselves what is a PL/SQL "sql_script" and if we want this grammar to be used with a specific interpreter or to be generic with any interpreter, in this case, what are the minimum established rules that should be common to any interpreter.

As it is defined:

```
sql_script
    ((unit_statement | sql_plus_command) SEMICOLON?)* EOF 
    ; 

unit_statement 
    : transaction_control_statements 
    | alter_cluster 
   ...
    ;

sql_plus_command
    : '/' 
    | EXIT 
    | PROMPT_MESSAGE 
    | SHOW (ERR | ERRORS) 
    | START_CMD 
    | whenever_command 
    | set_command 
    ;
```

Should we consider _SQL*Plus/SQLcl_ as the main PL/SQL interpreter? 

I think it is reasonable that the default tool in any Oracle installation such as _**SQL*Plus**_ should be considered as the main interpreter of PL/SQL. And therefore as one of our tools where valid sql script (according with grammar) should run without any problem. 

For _SQL*Plus_ the semicolon at the end of a line is considered as the end of the sentence. And the slash (solidus) in a new line is considered as the end of a sentence (if it does not end on semicolon) or as the end of a PL/SQL block since the semicolon is part of the PL/SQL definition.

This is valid:
` insert into A (ID,TEST) values ('0', 'test text');`

This is also valid (statement do not ends on semicolon):
```
insert into A (ID,TEST) values ('0', 'test text')
/
```

Slash (solidus) is also considered as "execute SQL buffer" and this is an ambiguity that should be solved.

This statemen execute the **insert** twice, once for statement ending on semicolon and once because of "slash" is consideder as "execute SQL buffer"!
```
insert into A (ID,TEST) values ('0', 'test text');
/
```

Should we consider _SQL Developer_ as the main interpreter? My opinion is no. It is an additional tool from Oracle. However, in _SQL Developer_ it is also necessary to indicate an end of sentence if you have several sentences togheter. You haven't got any _SQL buffer_ so slash is not interpreted on _SQL Developer_ as it is done on _SQL*Plus_.

Should we consider _JDBC/ODBC_ as an sql script interpreter? I think that as you have indicated, _JDBC/ODBC_ only works with statements (and PL/SQL blocks), so the grammatical rule "unit_statement" should be used not the "sql_script" one.

We should think therefore about how to consider which is the end of a sentence in an **sql script**. Both programs _SQL*Plus_ and _SQL Developer_ need some way to indicate when a DML and DDL sentence has finished.  Only _SQL*Plus_ requires the slash as end of PL/SQL Block.

Other third-party tools as DBeaver or Toad also requires that individual sentence to be separated by semicolon or slash when they are part of a script.

Therefore, in my opinion, single statements do not end on semicolon (it is not necesary); but several statements (aka sql script) should have individually an statemet separator usually a semicolon (for DML or DDL statements) or slash (for PL/SQL blocks).

**Comment by Pilo11** [comment]
Yes, this is correct. Then we should think about different grammar starting points (sql_script and sql_develoepr_script or something like this). 

> Therefore, in my opinion, single statements do not end on semicolon (it is not necessary)

Almost. If you want to create a package with a single statement "create ore replace package ... as begin ... end;".
This time the semicolon is needed! You cannot remove it without getting a syntax error (ODBC/JDBC).

> but several statements (aka sql script) should have individually an statemet separator usually a semicolon (for DML or DDL statements) or slash (for PL/SQL blocks).

This could be a solution.
